applications/OptimizationApplication/python_scripts/controls/shape/vertex_morphing_shape_control.py

```diff
@@ -0,0 +1,363 @@
+            else:
+                self.filter = KratosOA.NodalExplicitFilter(self.filter_model_part, self.parameters["filter_settings"]["filter_function_type"].GetString(), 1000)
+
+            filter_radius_field = Kratos.Expression.NodalExpression(self.filter_model_part)
+            Kratos.Expression.LiteralExpressionIO.SetData(filter_radius_field, self.parameters["filter_settings"]["radius"].GetDouble())
+            self.filter.SetFilterRadius(filter_radius_field)
+            self.control_field = Kratos.Expression.NodalExpression(self.filter_model_part)
+            Kratos.Expression.LiteralExpressionIO.SetData(self.control_field,[0,0,0])
+        else:
+            self.filter.Initialize()
+            self._SetFixedModelPartValues(False)
+            self.control_field = self.filter.UnFilterField(self.GePhysicalField())
+            self._SetFixedModelPartValues(True)
+
+        # initialize_mesh_motion
+        if self.has_mesh_motion:
+            self._mesh_moving_analysis.Initialize()
+
+        # now update
+        self._UpdateAndOutputFields(self.GetEmptyField())
+        self.is_initialized = True
+
+    def SetupMeshMotion(self):
+        self.has_mesh_motion = True
+        if "mesh_motion" in self.filter_type:
+            self.mesh_motion_params = GetMeshMotionParameters(self.parameters["mesh_motion_settings"],self.model_part_operation.GetRootModelPart().Name,self.model_part_operation.GetModelPartFullName())
+            self._mesh_moving_analysis = MeshMovingAnalysis(self.model, self.mesh_motion_params)
+        else:
+            self.has_mesh_motion = False
+
+    def Check(self) -> None:
+        if self.is_filter_implicit:
+            return self.filter.Check()
+
+    def Finalize(self) -> None:
+        if self.is_filter_implicit:
+            self.filter.Finalize()
+        if self.has_mesh_motion:
+            self._mesh_moving_analysis.Finalize()
+
+    def GetPhysicalKratosVariables(self) -> 'list[SupportedSensitivityFieldVariableTypes]':
+        return [KratosOA.SHAPE]
+
+    def GetEmptyField(self) -> ContainerExpressionTypes:
+        field = Kratos.Expression.NodalExpression(self.filter_model_part)
+        Kratos.Expression.LiteralExpressionIO.SetData(field, [0,0,0])
+        return field
+
+    def GetControlField(self) -> ContainerExpressionTypes:
+        return self.control_field
+
+    def GePhysicalField(self) -> ContainerExpressionTypes:
+        physical_shape_field = Kratos.Expression.NodalExpression(self.filter_model_part)
+        Kratos.Expression.NodalPositionExpressionIO.Read(physical_shape_field, Kratos.Configuration.Initial)
+        return physical_shape_field
+
+    def MapGradient(self, physical_gradient_variable_container_expression_map: 'dict[SupportedSensitivityFieldVariableTypes, ContainerExpressionTypes]') -> ContainerExpressionTypes:
```

Comment:
can you add please time decorator to expensive functions?

Reply:
Done.